### PR TITLE
004 add login system

### PR DESCRIPTION
**Pull request by jericoatos**

### CODE REVIEW

**Review by niall-mccloy — COMMENTED** [review]
just one comment

**Review by niall-mccloy — APPROVED** [review]

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        env:
+          FLYWAY_PASSWORD: ${{ secrets.DB_PASSWORD }}
+          FLYWAY_USER: ${{ secrets.DB_USERNAME }}
+          FLYWAY_URL: jdbc:mysql://${{ secrets.DB_HOST }}/${{ secrets.DB_NAME }}
+          FLYWAY_BASELINE_ON_MIGRATE: true
+        run: |
+          mvn flyway:clean
+          mvn flyway:migrate -Dflyway.baselineOnMigrate=true
+
+      - name: Maven build with Integration and Unit Tests
```
Comment (mlynch147) [inline .github/workflows/test.yml]: AFAIK - mvn test will only run unit tests so no env variables are needed here...

If you want to also run the integration tests, then you will have to have a step that runs... "mvn clean integration-test"

]

Comment (jericoatos) [reply .github/workflows/test.yml]: Resolved - mvn test runs unit and integration test

File: src/main/java/org/example/daos/AuthDao.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        try (Connection connection = DatabaseConnector.getConnection()) {
+            String query = "SELECT `email`, `password`, `roleId`"
+                    + "FROM `User`"
+                    + "WHERE `email` = ?;";
+
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setString(1, loginRequest.getEmail());
+
+            ResultSet resultSet = statement.executeQuery();
+            Argon2PasswordEncoder arg2SpringSecurity = getPasswordEncoder();
```
Comment (mlynch147) [inline src/main/java/org/example/daos/AuthDao.java]: No need to create this object if we have an empty result set... this cold be moved inside the while statement

File: src/main/java/org/example/daos/AuthDao.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            while (resultSet.next()) {
+                String encodedPassword =
+                        resultSet.getString("password");
+                if (
+                        arg2SpringSecurity.matches(
```
Comment (mlynch147) [inline src/main/java/org/example/daos/AuthDao.java:33]: does "matches" catch case differences?

File: src/main/java/org/example/util/PasswordEncoder.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package org.example.util;
+
+import org.springframework.security.crypto.argon2.Argon2PasswordEncoder;
+
+public final class PasswordEncoder {
+    private PasswordEncoder() { }
+
+    private static final Argon2PasswordEncoder ARGON_2_PASSWORD_ENCODER
+            = new Argon2PasswordEncoder(16, 32, 1, 60000, 10);
```
Comment (mlynch147) [inline src/main/java/org/example/util/PasswordEncoder.java:9]: Did checkstyle not complain about magic numbers here?

File: src/main/java/org/example/daos/AuthDao.java (hunk 3 above)
Comment (jericoatos) [reply src/main/java/org/example/daos/AuthDao.java:33]: Yep, matches catches case differences

File: src/main/resources/db/migration/V6__create_role_table.sql
Comment (mlynch147) [inline src/main/resources/db/migration/V6__create_role_table.sql:1]: can you fix the formatting on this file please... its very hard to read with just a single word on each line

File: src/main/resources/db/migration/V7__create_user_table.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS `User` (
+                                      `email` VARCHAR(64) PRIMARY KEY NOT NULL,
```
Comment (mlynch147) [inline src/main/resources/db/migration/V7__create_user_table.sql]: fix spacing please

File: src/test/java/org/example/validators/EmailValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+public class EmailValidatorTest {
+
+    @Test
+    public void isValidEmail_shouldReturnTrue_forValidEmail() {
+        String VALID_EMAIL = "valid.address@gmail.com";
+        assertTrue(isValidEmail(VALID_EMAIL));
+    }
+
+    @Test
+    public void isValidEmail_shouldReturnFalse_forInvalidEmail() {
```
Comment (mlynch147) [inline src/test/java/org/example/validators/EmailValidatorTest.java:18]: Can you please add some more invalid tests as there could be a number of different things that make an email invalid

File: src/main/java/org/example/util/PasswordEncoder.java (hunk 4 above)
Comment (jericoatos) [reply src/main/java/org/example/util/PasswordEncoder.java:9]: It didn't complain. When I tried to change the numbers eg. 'private final int saltLength = 16;' it wouldn't accept it.

File: src/test/java/org/example/validators/EmailValidatorTest.java (hunk 6 above)
Comment (mlynch147) [reply src/test/java/org/example/validators/EmailValidatorTest.java:18]: great job

**Review by mlynch147 — APPROVED** [review]

**Comment by cormaco7** [comment]
added selenium, related dependencies and created LoginTest - needs reviewed before the rest of the testing can be done.

**Review by orlaghmccloskey1 — CHANGES_REQUESTED** [review]
can't use xpath - need to find by id

**Review by orlaghmccloskey1 — COMMENTED** [review]
Merge conflicts

**Review by orlaghmccloskey1 — APPROVED** [review]

**Pull request merged**